text/0000-custom-model-classes.md

```diff
@@ -0,0 +1,158 @@
+
+We will need four separate changes: 
+
+- A way to instantiate custom Records and have those records be notified of changes in Record Data
+- A way to access schema information not from the Record Class
+- Replacement of few store methods that take DS.Models
+- Deprecate passing DS.Model classes to adapter/serializers/exposing them on Snapshots
+
+## Instantiating custom Records
+
+First we need to expose a method on the store for instantiating a record, and way for the record to get notified when RecordData values change:
+
+```ts
+interface RecordDataWrapper { // proxies a subset of RD methods and hides the rest
+  performMutation(operation: Operation): void
+      
+  getAttr(identifier, key)
+  isAttrDirty(identifier, key)
+    
+  changedAttributes(identifier)
+  hasChangedAttributes(identifier)
+  rollbackAttributes(identifier)
+  
+  getRelationship(identifier, key)
+  
+  setRecordId(identifier, id)
+  
+  getErrors(identifier)
+  getMeta(identifier)
+  
+  isNew(identifier)
+  isDeleted(identifier)
+}
+  
+interface RecordDataFor {
+  (identifier: RecordIdentifier): RecordDataWrapper
+}
+    
+interface NotificationCallback {
+  (identifier: Identifier, notificationType: 'attributes' | 'relationships' | 'errors'  | 'meta'  | 'unload') => void;
+}
+    
+interface NotificationManager {
+  subscribeAttributeChanges(identifier, NotificationCallback)
+  subscribeRelationshipChanges(identifier, NotificationCallback)
+  subscribeErrorsChanges(identifier, NotificationCallback)
+  subscribeMetaChanges(identifier, NotificationCallback)
+  subscribeUnload(identifier, NotificationCallback)
+  unsubscribe(identifier, NotificationCallback)
+ }
+    
+class Store {
+  instantiateRecord(
+    identifier: RecordIdentifier, 
+    createArgs: CreateArgs, 
+    recordDataFor: RecordDataFor, 
+    notificationManager: NotificationManager): unknown)
+}
+```
+
+Instead of exposing the entire RecordData Class, we expose only the local facing methods and hide the server/adapter facing methods that the Model should not have access to, by giving a RecordData lookup method to `instantiateRecord`
+
+## Exposing schema information
+
+We currently keep schema information on `DS.Model` class. In order to allow for custom Model implementations we need to allow lookup of schema info from the store. We already have specified a schema api that RecordData consumes: `attributesDefinitionFor` and `relationshipDefinitionFor`. We would define these methods publicly on the store for addons to use. **These methods are not ergonomic on purpose.** They match the current Record Data apis and are designed as a stepping stone on the way of having a better, user facing schema APIs. The schema info is currently primarily geared towards being used by the internal relationship handling layer and the serializer/adapter layers.
+
+```ts
+export interface RelationshipSchema {
+  kind: 'hasMany'| 'belongsTo';
+  type: string;
+  options: { [key: string]: any } ;
+  name: string;
```

Comment:
we should validate that this is in fact what we currently expose / want to document as info concerning `inversePropertyName` `inverseType` and `isPolymorphic` is typically available as well.

Reply:
Thats a very good point. `isPolymorphic` should be available from the options. We currently in practice expose a bunch of underscored properties for inverses, but all of that data is calculable given the exposed public data. I would prefer to keep it private, and figure out in the schema work if we need to expose it